analytics/tracker.go

```diff
@@ -26,110 +78,166 @@ type Tracker interface {
-	inputBytes, err := json.Marshal(inputs)
-	if err != nil {
-		log.Errorf("Failed to marshal inputs: %s", err)
+	if !t.stateChecker.Enabled() {
+		return
+	}
+
+	extraProperties := []analytics.Properties{properties, prepareStartProperties(infoModel)}
+	if len(inputs) > 0 {
+		inputBytes, err := json.Marshal(inputs)
+		if err != nil {
+			t.SendCLIWarning(fmt.Sprintf("Failed to marshal inputs: %s", err.Error()))
+		} else {
+			extraProperties = append(extraProperties, analytics.Properties{inputsProperty: string(inputBytes)})
+		}
 	}
-	t.tracker.Enqueue("step_started", properties, prepareStartProperties(infoModel), analytics.Properties{"inputs": string(inputBytes)})
+
+	t.tracker.Enqueue(stepStartedEventName, extraProperties...)
 }
 
 // SendStepFinishedEvent ...
 func (t tracker) SendStepFinishedEvent(properties analytics.Properties, results models.StepRunResultsModel) {
+	if !t.stateChecker.Enabled() {
+		return
+	}
+
+	var eventName string
+	var extraProperties analytics.Properties
+
 	switch results.Status {
 	case models.StepRunStatusCodeSuccess:
-		t.tracker.Enqueue("step_finished", properties, analytics.Properties{"status": "successful"})
+		eventName = stepFinishedEventName
+		extraProperties = analytics.Properties{statusProperty: successfulValue}
 		break
 	case models.StepRunStatusCodeFailed, models.StepRunStatusCodeFailedSkippable:
-		failedProperties := analytics.Properties{"status": "failed"}
-		failedProperties.AppendIfNotEmpty("error_message", results.ErrorStr)
-		t.tracker.Enqueue("step_finished", properties, failedProperties)
+		eventName = stepFinishedEventName
+		extraProperties = analytics.Properties{statusProperty: failedValue}
+		extraProperties.AppendIfNotEmpty(errorMessageProperty, results.ErrorStr)
 		break
 	case models.StepRunStatusCodePreparationFailed:
-		failedProperties := prepareStartProperties(results.StepInfo)
-		failedProperties.AppendIfNotEmpty("error_message", results.ErrorStr)
-		t.tracker.Enqueue("step_preparation_failed", properties, failedProperties)
+		eventName = stepPreparationFailedEventName
+		extraProperties = prepareStartProperties(results.StepInfo)
+		extraProperties.AppendIfNotEmpty("error_message", results.ErrorStr)
 	case models.StepRunStatusCodeSkipped, models.StepRunStatusCodeSkippedWithRunIf:
-		startProperties := prepareStartProperties(results.StepInfo)
+		eventName = stepSkippedEventName
+		extraProperties = prepareStartProperties(results.StepInfo)
 		if results.Status == models.StepRunStatusCodeSkipped {
-			startProperties["reason"] = "build_failed"
+			extraProperties[reasonProperty] = buildFailedValue
 		} else {
-			startProperties["reason"] = "run_if"
+			extraProperties[reasonProperty] = runIfValue
 		}
-		t.tracker.Enqueue("step_skipped", properties, startProperties)
 	default:
-		panic("Unknown step status code")
+		panic(fmt.Sprintf("Unknown step status code: %d", results.Status))
```

Comment:
Should we ever panic in the analytics code? 🤔 Not sure what is worse, silent errors and malformed data, or crashing the Bitrise CLI.

Reply:
Maybe `SendCLIWarning()` would be a better alternative here.